neuralmonkey/decoders/decoder.py

```diff
@@ -185,60 +221,72 @@ def _encoder_projection(self, encoded_states):
         return tf.tanh(tf.matmul(dropped_input, weights) + biases)
 
 
-    def _dropout(self, var):
-        """Perform dropout on a variable
+    def _create_initial_state(self):
+        """Construct the part of the computation graph that computes the initial
+        state of the decoder."""
 
-        Arguments:
-            var: The variable to perform the dropout on
-        """
-        return tf.nn.dropout(var, self.dropout_placeholder)
+        if len(self.encoders) == 0:
+            self.initial_state = tf.zeros([self.rnn_size])
+        else:
+            encoders_out = tf.concat(1, [e.encoded for e in self.encoders])
+            if self.project_encoder_outputs:
+                encoders_out = self._encoder_projection(encoders_out)
+
+            self.initial_state = self._dropout(encoders_out)
+
+        ## Broadcast the initial state to the whole batch if needed
+        ## CHANGE: REFACTORED + EXPAND DIMS USED INSTEAD OF [:1]. ALSO, THE
+        ## DIMENSIONS ARE DIFFERENT
+        if len(self.initial_state.get_shape()) == 1:
+            assert self.initial_state.get_shape()[0].value == self.rnn_size
+
+            batch_size = tf.shape(self.go_symbols)[1]
+            tiles = tf.tile(self.initial_state, tf.expand_dims(batch_size, 0))
+
+            self.initial_state = tf.reshape(tiles, [-1, self.rnn_size])
 
 
-    def _input_embeddings(self):
+    def _create_embedding_matrix(self):
         """Create variables and operations for embedding of input words
 
-        If we are reusing word embeddings, this function takes
-        them from the first encoder
+        If we are reusing word embeddings, this function takes the embedding
+        matrix from the first encoder
         """
-        if self.reuse_word_embeddings:
-            return self.encoders[0].word_embeddings
-
         # NOTE In the Bahdanau paper, they say they initialized some weights
         # as orthogonal matrices, some by sampling from gauss distro with
         # stddev=0.001 and all other weight matrices as gaussian with
         # stddev=0.01. Embeddings were not among any of the special cases so
         # I assume that they initialized them as any other weight matrix.
-        return tf.get_variable(
-            "word_embeddings", [self.vocabulary_size, self.embedding_size],
-            initializer=tf.random_normal_initializer(stddev=0.01))
 
+        if self.reuse_word_embeddings:
+            self.embedding_matrix = self.encoders[0].word_embeddings
+        else:
+            self.embedding_matrix = tf.get_variable(
+                "word_embeddings", [self.vocabulary_size, self.embedding_size],
+                initializer=tf.random_normal_initializer(stddev=0.01))
 
-    def _training_placeholders(self):
-        """Defines data placeholders for training the decoder"""
 
-        inputs = [tf.placeholder(tf.int64, [None], name="decoder_{}".format(i))
-                  for i in range(self.max_output + 2)]
+    def _embed(self, inputs):
```

Comment:
To bych přejmenoval na `_embed_and_dropout`, ať to křičí z názvu i bez docstringu.